services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/NestedAttributeName.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.enhanced.dynamodb;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+
+/**
+ * High-level representation of a DynamoDB 'NestedAttributeName' that can be used in various situations where the API requires
+ * or accepts an Nested Attribute Name.
+ * Simple Attribute Name can be represented by passing just the name of the attribute.
+ * Nested Attributes are represented by List of String where each index of list corresponds to Nesting level Names.
+ * <p> While using attributeToProject in {@link software.amazon.awssdk.enhanced.dynamodb.model.QueryEnhancedRequest}
+ * and {@link software.amazon.awssdk.enhanced.dynamodb.model.ScanEnhancedRequest} we need  way to represent Nested Attributes.
+ * The normal DOT(.) separator is not recognized as a Nesting level separator by BynamoDB request,
+ * thus we need to use NestedAttributeName
+ * which can be used to represent Nested attributes.
+ * <p> Example : NestedAttributeName.create("foo") corresponds to a NestedAttributeName with elements list
+ * with single element foo which represents Simple attribute name "foo" without nesting.
+ * <p>NestedAttributeName.create("foo", "bar") corresponds to a NestedAttributeName with elements list "foo", "bar"
+ * respresenting nested attribute name "foo.bar".
+ */
+@SdkPublicApi
+public final class NestedAttributeName {
+
+    private final List<String> elements;
+
+    private NestedAttributeName(List<String> nestedAttributeNames) {
+        this.elements = nestedAttributeNames != null
+                ? Collections.unmodifiableList(nestedAttributeNames) : null;
```

Comment:
Having null elements should be invalid. Can we add a validator here?

Reply:
Added for Null as well as Empty list